drivers/generic_gpio_common.c

```diff
@@ -0,0 +1,444 @@
+	upsdebugx(LOG_DEBUG, "rules =[%s]", rulesString);
+	while((lexType=get_rule_lex(rulesString, &startPos, &endPos))>0 && lexStatus>=0) {
+		memset(lexBuff, 0, sizeof(lexBuff));
+		strncpy(lexBuff, (char *)(rulesString+startPos), endPos-startPos);
+		upsdebugx(
+			LOG_DEBUG,
+			"rules start %d, end %d, lexType %d, lex [%s]",
+			startPos,
+			endPos,
+			lexType,
+			lexBuff
+		);
+		switch(lexStatus) {
+			case 0:
+				if(lexType!='a') {
+					lexStatus=-1;
+				} else {
+					lexStatus=1;
+					upsfd->rulesCount++;
+					upsfd->rules=xrealloc(upsfd->rules, (size_t)(sizeof(upsfd->rules[0])*upsfd->rulesCount));
+					upsfd->rules[upsfd->rulesCount-1]=xcalloc(sizeof(rulesint), 1);
+					strncpy(upsfd->rules[upsfd->rulesCount-1]->stateName, (char *)(rulesString+startPos), endPos-startPos);
+					upsfd->rules[upsfd->rulesCount-1]->stateName[endPos-startPos]=0;
+				}
+			break;
+			case 1:
+				if(lexType!='=') {
+					lexStatus=-1;
+				} else {
+					lexStatus=2;
+				}
+			break;
+			case 2:
+				if(lexType=='^') {
+					lexStatus=3;
+					add_rule_item(upsfd, RULES_CMD_NOT);
+				} else if(lexType=='0') {
+					lexStatus=4;
+					add_rule_item(upsfd, atoi((char *)(rulesString+startPos)));
+				} else {
+					lexStatus=-1;
+				}
+			break;
+			case 3:
+				if(lexType!='0') {
+					lexStatus=-1;
+				} else {
+					lexStatus=4;
+					add_rule_item(upsfd, atoi((char *)(rulesString+startPos)));
+				}
+			break;
+			case 4:
+				if(lexType=='&') {
+					lexStatus=2;
+					add_rule_item(upsfd, RULES_CMD_AND);
+				} else if(lexType=='|') {
+					lexStatus=2;
+					add_rule_item(upsfd, RULES_CMD_OR);
+				}
+				else if(lexType==';') {
+					lexStatus=0;
+				} else {
+					lexStatus=-1;
+				}
+			break;
+			default:
+				lexStatus=-1;
+			break;
+		}
+		if(lexStatus==-1)
+			fatalx(LOG_ERR, "Line processing rule error at position %d", startPos);
+		startPos=endPos;
+	}
+
+	upsdebugx(LOG_DEBUG, "rules count [%d]", upsfd->rulesCount);
+	for(i=0; i<upsfd->rulesCount; i++) {
+		upsdebugx(
+			LOG_DEBUG,
+			"rule state name [%s], subcount %d",
+			upsfd->rules[i]->stateName,
+			upsfd->rules[i]->subCount
+		);
+		for(j=0; j<upsfd->rules[i]->subCount; j++) {
+			upsdebugx(
+				LOG_DEBUG,
+				"[%s] substate %d [%d]",
+				upsfd->rules[i]->stateName,
+				j,
+				upsfd->rules[i]->cRules[j]
+			);
+		}
+	}
+
+	upsfd->upsLinesCount=0;
+	upsfd->upsMaxLine=0;
+	for(i=0; i<upsfd->rulesCount; i++) {
+		for(j=0; j<upsfd->rules[i]->subCount; j++) {
+			int pinOnList=0;
+			for(k=0; k<upsfd->upsLinesCount && !pinOnList; k++) {
+				if(upsfd->upsLines[k]==upsfd->rules[i]->cRules[j]) {
+					pinOnList=1;
+				}
+			}
+			if(!pinOnList) {
+				if(upsfd->rules[i]->cRules[j]>=0) {
+					upsfd->upsLinesCount++;
+					upsfd->upsLines=xrealloc(upsfd->upsLines, sizeof(upsfd->upsLines[0])*upsfd->upsLinesCount);
+					upsfd->upsLines[upsfd->upsLinesCount-1]=upsfd->rules[i]->cRules[j];
+					if(upsfd->upsLines[upsfd->upsLinesCount-1]>upsfd->upsMaxLine) {
```

Comment:
General coding style (all around the file) - this would be much more readable with spaces around operators like `<`, `>`, `=`, `==` etc., and we have spaces between `if/for` and the `(` parenthesis after it :)